internal/exec/stages/files/units.go

```diff
@@ -64,10 +92,69 @@ func (s *stage) createUnits(config types.Config) error {
 			s.relabel(relabelpath)
 		}
 	}
-	// and relabel the preset file itself if we enabled/disabled something
-	if enabledOneUnit {
-		s.relabel(util.PresetPath)
+	// if we have presets then create the systemd preset file.
+	if len(presets) != 0 {
+		if err := s.createSystemdPresetFile(presets); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// parseInstanceUnit is a helper function to extract the name
+// and coresponding instances for a given instantiated unit.
+func parseInstanceUnit(unit types.Unit) (string, string, error) {
+	at := strings.Index(unit.Name, "@")
+	dot := strings.LastIndex(unit.Name, ".")
+	if dot == -1 {
+		return "", "", errors.ErrNoDotSuffix
+	}
+	instance := unit.Name[at+1 : dot]
+	serviceInstance := unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]
+	return serviceInstance, instance, nil
+}
+
+// createSystemdPresetFile creates the presetfile for enabled/disabled
+// instantiated/systemd units.
+func (s *stage) createSystemdPresetFile(presets map[string]*Preset) error {
+	hasInstanceUnit := false
+	for _, value := range presets {
```

Comment:
`hasInstanceUnit` isn't used until after the second loop, so we don't need a separate loop for this.

Reply:
How about adding this variable inside the second for loop?
```
if value.instantiatable {
                       hasInstanceUnit = true
			unitString = fmt.Sprintf("%s %s", unitString, strings.Join(value.instances, " "))
}
```